proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,139 @@
+- Start Date: 2022-10-25
+- Reference Issues: TODO
+- Implementation PR: <!-- leave empty -->
+
+# Summary
+
+Currently, Astro's [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) enables developers to switch between `'static'` (default) and `'server'` outputs. One of our most requested features is to enable route-level control of `output` so that some routes can be server-rendered while others are statically generated.
+
+# Example
+
+This RFC proposes a new, opt-in `output` mode named `'hybrid'`.
```

Comment:
```
output: 'hybrid'
```

I'd strongly push for this to remain a feature of output: 'server'. Basically, that we don't introduce a new third mode but instead add this new `export const output` support to anyone with `output: 'server'` enabled, that will let you prerender pages. 

Conceptually, this makes more sense to me. I'd always seen `server` vs. `static` more as a boolean: "does this build output include a server or not?" There are huge differences between the two that we need to build for and that you need to account for as a user deploying your site. But, this feature doesn't really change that story.

As both an implementation and a user experience, I think this makes more sense as a feature of `output: server` vs. a thing that is different from it. As I understand the RFC, the `hybrid` build output and deploy story are both is identical to `server`.

Reply:
Implementation-wise, there are a lot of reasons to consider shipping `export const output` support as a feature of `output: 'server'`:

**DX (Defaults):** By making this an opt-in story for `output: 'server'`, you solve for all of the drawbacks and complexities in this RFC related to default behavior concerns and migration strategy. I think this is a deeply tricky part of the current RFC that doesn't have a good answer, since "hybrid" doesn't tell you the user's default intention as much as "static" and "server" do. Because of that, some group of users will always be unhappy with the "hybrid" defaults not being what they expected.

**DX (API Simplicity):** Introducing a third mode requires more documentation and will lead to more user questions. When do I want to use hybrid mode? How does X work in hybrid mode? How does Y work in hybrid mode? I'm having problem Z, does switching to hybrid mode solve that?

**Maintenance:** I think its fair to say that having to support two different top-level "modes" is one of our biggest maintenance overheads today. Ex: Anytime we ship a new feature, we need to 2x our testing to make sure it works consistently across "static" and "server". Anytime a user reports a bug, we need to make sure we know the `output` mode in case the behavior changes across modes, etc. Adding a third top-level option to support this single `export const output` feature would increase this complexity.